From b1c02054b29cf7acc429e61699602c7808cc2e44 Mon Sep 17 00:00:00 2001
From: Don <don.j.olmstead@gmail.com>
Date: Mon, 23 Apr 2018 18:49:42 -0700
Subject: [PATCH] Add declspec for exporting as a shared library

This adds a default declaration of __declspec to WEBP_EXTERN. Also removes an erroneous WEBP_EXTERN in dsp.h.
---
 src/dsp/dsp.h    | 2 +-
 src/webp/types.h | 8 +++++++-
 2 files changed, 8 insertions(+), 2 deletions(-)

diff --git a/src/dsp/dsp.h b/src/dsp/dsp.h
index 298c721a..96962cb0 100644
--- a/src/dsp/dsp.h
+++ b/src/dsp/dsp.h
@@ -225,7 +225,7 @@ typedef enum {
 } CPUFeature;
 // returns true if the CPU supports the feature.
 typedef int (*VP8CPUInfo)(CPUFeature feature);
-WEBP_EXTERN VP8CPUInfo VP8GetCPUInfo;
+extern VP8CPUInfo VP8GetCPUInfo;
 
 //------------------------------------------------------------------------------
 // Init stub generator
diff --git a/src/webp/types.h b/src/webp/types.h
index 47f7f2b0..b53e67d6 100644
--- a/src/webp/types.h
+++ b/src/webp/types.h
@@ -36,10 +36,16 @@ typedef long long int int64_t;
 #define WEBP_INLINE __forceinline
 #endif  /* _MSC_VER */
 
+#ifndef __has_declspec_attribute
+#define __has_declspec_attribute(x) 0
+#endif
+
 #ifndef WEBP_EXTERN
 // This explicitly marks library functions and allows for changing the
 // signature for e.g., Windows DLL builds.
-# if defined(__GNUC__) && __GNUC__ >= 4
+# if defined(WIN32) || (__has_declspec_attribute(dllexport) && __has_declspec_attribute(dllimport))
+#  define WEBP_EXTERN __declspec(dllexport)
+# elif defined(__GNUC__) && __GNUC__ >= 4
 #  define WEBP_EXTERN extern __attribute__ ((visibility ("default")))
 # else
 #  define WEBP_EXTERN extern
-- 
2.29.2.windows.1

